kratos/mpi/utilities/pointer_communicator.h

```diff
@@ -0,0 +1,373 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Riccardo Rossi
+//
+
+
+#if !defined(KRATOS_POINTER_COMMUNICATOR_H_INCLUDED )
+#define  KRATOS_POINTER_COMMUNICATOR_H_INCLUDED
+
+
+// System includes
+#include <string>
+#include <iostream>
+
+
+// External includes
+
+
+// Project includes
+#include "includes/define.h"
+#include "includes/data_communicator.h"
+#include "includes/global_pointer.h"
+#include "includes/mpi_serializer.h"
+#include "containers/global_pointers_vector.h"
+#include "containers/global_pointers_unordered_map.h"
+
+#include "mpi/utilities/mpi_coloring_utilities.h"
+
+namespace Kratos
+{
+///@addtogroup ApplicationNameApplication
+///@{
+
+///@name Kratos Globals
+///@{
+
+///@}
+///@name Type Definitions
+///@{
+
+///@}
+///@name  Enum's
+///@{
+
+///@}
+///@name  Functions
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+template< class TPointerDataType, class TSendType >
+class ResultsProxy
+{
+public:
+
+    ResultsProxy(
+        int current_rank,
+        GlobalPointersUnorderedMap< TPointerDataType, TSendType > NonLocalData,
+        std::function< TSendType(GlobalPointer<TPointerDataType>&) > user_function
+    ):
+        mcurrent_rank(current_rank), mNonLocalData(NonLocalData), muser_function(user_function)
+    {}
+
+    /// Destructor.
+    virtual ~ResultsProxy() {}
+
+    /**this function returns the effect of "user_function(gp)" both if the gp is locally owned
+     * and if it is remotely owned
+     */
+    TSendType Get(GlobalPointer<TPointerDataType>& gp)
+    {
+        if(gp.GetRank() == mcurrent_rank)
+            return muser_function(gp);
+        else
+            return mNonLocalData[gp];
+    }
+
+private:
+    const int mcurrent_rank;
+    GlobalPointersUnorderedMap< TPointerDataType, TSendType > mNonLocalData;
+    std::function< TSendType(GlobalPointer<TPointerDataType>&) > muser_function;
```

Comment:
The format here is inconsistent. I assume it should be `mCurrentRank`, `mNonLocalData`, `mUserFunction`.

Reply:
done